temporal-serviceclient/src/main/java/io/temporal/internal/retryer/GrpcAsyncRetryer.java

```diff
@@ -40,25 +35,34 @@ class GrpcAsyncRetryer<R> {
   private static final Logger log = LoggerFactory.getLogger(GrpcRetryer.class);
 
   private final ScheduledExecutorService executor;
-  private final GrpcRetryer.GrpcRetryerOptions options;
-  private final GetSystemInfoResponse.Capabilities serverCapabilities;
+  private final Supplier<GetSystemInfoResponse.Capabilities> serverCapabilities;
```

Comment:
Will let @Quinn-With-Two-Ns review as he is more familiar w/ the Java side, but I have a couple of questions:

* Can you confirm (sorry I'm lazy) that this server capabilities getter is memoized? Meaning after the first time it is fetched, it never fetches again?
* Go does this similar approach for lazy server capabilities (which is also our first-call health check), but in order to validate the connection on worker start, we eagerly make sure the this capabilities call is made on worker start. Can we make sure Java does the same? Unsure of behavior today, but in other SDKs we've seen lazy clients w/ bad connection options just fail inside the worker poller never to have user-facing error.

Reply:
Regarding point 1, the Supplier comes from [`ChannelManager.getServerCapabilities()`](https://github.com/temporalio/sdk-java/blob/master/temporal-serviceclient/src/main/java/io/temporal/serviceclient/ChannelManager.java#L325) which seems to be using `j.u.c.CompletableFuture` as the world's strangest memoization cache.

Reply:
:+1: Now just need to confirm we're doing some form of eager connection and capability lookup on worker start. I would guess we are.

Reply:
Okay, I've refactored a bunch of stuff to use eager `Capabilities` messages.  I also swapped out `CompletableFuture` for an `AtomicReference`, which it probably should have been using all along.

Reply:
Hrmm, I can see that for the final answer, but completable future can help making sure an invocation is lazy.

Reply:
I'm not sure I understand what you mean?  1) I thought making invocations lazy is exactly what you're asking us *not* to do, and 2) the future isn't exposed to the caller in the existing code, we just call `getNow()` on it when we want to access it.

Reply:
> I thought making invocations lazy is exactly what you're asking us not to do

Only at worker start, not in general. Sorry if I wasn't clear. Today's Java devs (unfortunately) expect creation of the client to not make any outbound calls IIUC.

Reply:
Ah.  So it turns out this *does* currently get called on connect, since `ChannelManager.connect` calls `ChannelManager.healthCheck` in a `retryWithResult`.  (Unfortunately, `healthCheck` is doing a gRPC `Health.check` and not a `WorkflowService.getSystemInfo`, but that's a different problem.)

I'm currently looking into ways I can call WS.gSI from ChannelManager.connect to populate the `AtomicReference`, which would allow `getServerCapabilities` to be implemented as `return capabilitiesAtomicReference::get`.